packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -0,0 +1,387 @@
+
+        OptimismMintableERC20Factory(impl).initialize(address(0));
+
+        OptimismMintableERC20Factory(Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY).initialize(
+            Predeploys.L2_STANDARD_BRIDGE
+        );
+
+        _checkOptimismMintableERC20Factory(impl);
+    }
+
+    /// @notice This predeploy is following the saftey invariant #1.
+    ///         This contract has no initializer.
+    function _setL1BlockNumber() internal {
+        _setImplementationCode(Predeploys.L1_BLOCK_NUMBER, "L1BlockNumber");
+    }
+
+    /// @notice This predeploy is following the saftey invariant #1.
+    ///         This contract has no initializer.
+    function _setGasPriceOracle() internal {
+        _setImplementationCode(Predeploys.GAS_PRICE_ORACLE, "GasPriceOracle");
+    }
+
+    /// @notice This predeploy is following the saftey invariant #3.
+    function _setGovernanceToken() internal {
+        if (!cfg.enableGovernance()) {
+            console.log("Governance not enabled, skipping setting governanace token");
+            return;
+        }
+
+        GovernanceToken token = new GovernanceToken();
+        console.log("Setting %s implementation at: %s", "GovernanceToken", Predeploys.GOVERNANCE_TOKEN);
+        vm.etch(Predeploys.GOVERNANCE_TOKEN, address(token).code);
+
+        bytes32 _nameSlot = hex"0000000000000000000000000000000000000000000000000000000000000003";
+        bytes32 _symbolSlot = hex"0000000000000000000000000000000000000000000000000000000000000004";
+        bytes32 _ownerSlot = hex"000000000000000000000000000000000000000000000000000000000000000a";
+
+        vm.store(Predeploys.GOVERNANCE_TOKEN, _nameSlot, vm.load(address(token), _nameSlot));
+        vm.store(Predeploys.GOVERNANCE_TOKEN, _symbolSlot, vm.load(address(token), _symbolSlot));
+        vm.store(Predeploys.GOVERNANCE_TOKEN, _ownerSlot, bytes32(uint256(uint160(cfg.governanceTokenOwner()))));
+
+        /// Reset so its not included state dump
+        vm.etch(address(token), "");
+        vm.resetNonce(address(token));
+    }
+
+    /// @notice This predeploy is following the saftey invariant #1.
+    ///         This contract has no initializer.
+    /// @dev Previously the initial L1 attributes was set at genesis, to simplify,
+    ///      they no longer are so the resulting storage slots are no longer set.
+    function _setL1Block() internal {
+        _setImplementationCode(Predeploys.L1_BLOCK_ATTRIBUTES, "L1Block");
+    }
+
+    /// @dev Returns true if the adress is not proxied.
+    function _notProxied(address _addr) internal pure returns (bool) {
+        return _addr == Predeploys.GOVERNANCE_TOKEN || _addr == Predeploys.WETH9;
+    }
+
+    /// @dev Returns true if the address is a predeploy.
+    function _isDefinedPredeploy(address _addr) internal pure returns (bool) {
+        return _addr == Predeploys.L2_TO_L1_MESSAGE_PASSER || _addr == Predeploys.L2_CROSS_DOMAIN_MESSENGER
```

Comment:
There are 19 checks here but 32 addresses in `Predeploys.sol`. `Predeploys.sol` has 11 preinstalls that probably shouldn't be present in that lib, so that leaves 2 unaccounted for here

Reply:
> Predeploys.sol has 11 preinstalls that probably shouldn't be present in that lib...

Just FYI, this is addressed in #9317

Reply:
The two missing address are:

- `Predeploys.L1_MESSAGE_SENDER`
- `Predeploys.LEGACY_ERC20_ETH`

both of which are deprecated. However, `Predeploys.L1_MESSAGE_SENDER` is still included in the genesis alloc at the proxy address:

```json
"4200000000000000000000000000000000000001": {
      "code": "0x....",
      "storage": {
        "0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103": "0x0000000000000000000000004200000000000000000000000000000000000018"
      },
      "balance": "0x0"
    },
```

but no implementation for it is set as seen by the storage slot `0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc` not being set, and no `0xc0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d3c0d30001` alloc.

So nothing needs to happen, it's expected that there's only 19 checks in `_isDefinedPredeploy` because the other two address found in `Predeploys.sol` are deprecated

Reply:
As long as @tynes confirms both of these predeploys no longer need to be placed I'm good with resolving this thread. I know there are some deprecated contracts that are still referenced in some places, which is why I don't feel comfortable resolving this myself

Reply:
We no longer have the source code for the `L1MessageSender` contract in the repo. We should not set the implementation slot to anything in the proxy address `4200000000000000000000000000000000000001` and there should be nothing in the corresponding code address

Reply:
We have the legacy erc20 eth contract source but we do not set it at all or reference it in the Go code. I recommend that we delete the source code to save a little bit of compile time. It will only be set in OP Mainnet, I think its fine to leave the constant in the predeploys solidity file to ensure there isnt' a collision